akka-actor/src/main/scala/akka/dispatch/BatchingExecutor.scala

```diff
@@ -89,21 +89,22 @@ private[akka] trait BatchingExecutor extends Executor {
         parentBlockContext = prevBlockContext
         try processBatch(this) catch {
           case t: Throwable ⇒
+            parentBlockContext = null
             resubmitUnbatched()
             throw t
-        } finally {
-          _tasksLocal.remove()
-          parentBlockContext = null
-        }
+        } finally _tasksLocal.remove()
+        parentBlockContext = null
       }
     }
 
     override def blockOn[T](thunk: ⇒ T)(implicit permission: CanAwait): T = {
+      // need to rescue the parent BC before re-running this Batch
+      val parent = parentBlockContext
```

Comment:
Here I disagree: dereferencing `parent` clearly documents the intent that it is not assumed to be `null`, the code cannot possibly work without this, and the intended program semantics are that this is always non-`null`. We have countless uses of variables in our codebase that do not `require(x ne null)` before dereferencing `x`, why is this one different?